### PR TITLE
android sdks mapped

### LINKED ISSUE #15: sdk version info
Android SDK (sdkmanager --list)
iOS SDK (xcodebuild -showsdks)

### PR DESCRIPTION
**Pull request by GantMan**
Closes #15 

Keep in mind I have MANY Android SDKs installed

image

### CODE REVIEW

File: src/envinfo.js [hunk 1]
```diff
@@ -19,7 +19,7 @@ module.exports.helpers = helpers;
 module.exports.envinfo = function print(options) {
   const props = {
     System: ['OS', 'CPU', 'Free Memory', 'Total Memory', 'Shell'],
-    Binaries: ['Node', 'Yarn', 'npm', 'Watchman', 'Docker', 'Homebrew'],
+    Binaries: ['Node', 'Yarn', 'npm', 'Watchman', 'Docker', 'Homebrew', 'Android SDKs'],
```
Comment (tabrindle) [inline src/envinfo.js]: I think we should probably create a new top level for this.

File: src/helpers.js [hunk 2]
```diff
@@ -53,6 +53,29 @@ function getDarwinApplicationVersion(bundleIdentifier) {
...
+      androidSDKs.push(matcher[1]);
+    }
+  } catch(err) {
+    androidSDKs = ['Unknown'];
+  }
+
+  return androidSDKs;
+}
+
+function prettifyGetAllAndroidSDKs() {
```
Comment (tabrindle) [inline src/helpers.js]: could we use any of the formatters for this? It occurs to me we don't have any helpers that return an array of versions yet.

File: src/helpers.js [hunk 3]
```diff
@@ -53,6 +53,29 @@ function getDarwinApplicationVersion(bundleIdentifier) {
   return 'N/A';
 }
 
+function getAllAndroidSDKs() {
+  var androidSDKs = [];
+  try {
+    var installed = utils
+    .run(
+      '$ANDROID_HOME/tools/bin/sdkmanager --list'
```
Comment (tabrindle) [inline src/helpers.js]: This doesn't actually work on my system - for whatever reason (probably because I dont use it) ANDROID_HOME isnt defined. However, I do have sdkmanager somewhere in my path, as `which sdkmanager` works, and returns `/usr/local/bin/sdkmanager`

Thoughts?

**Comment by GantMan** [comment]
My bad!  I forgot semicolons.

Due to the need for iterating over while + calling exec (I hate the way JS does Regexp) I needed to bend one eslint rule.

File: src/helpers.js (hunk 3 above)
Comment (GantMan) [reply src/helpers.js]: So you could have multiple SDK managers.   Maybe we try to run the one you've identified as $ANDROID_HOME, and if that fails, we try to run the other? and if that fails, we call final failure?

File: src/envinfo.js (hunk 1 above)
Comment (GantMan) [reply src/envinfo.js]: What do you want to name it?

Comment (GantMan) [reply src/envinfo.js]: Just SDKs sounds good to me.

File: src/helpers.js (hunk 2 above)
Comment (GantMan) [reply src/helpers.js]: The reason I have it in a separate function is bc that's what I'll use, and I'll want the Array that is returned.

Maybe we can just have a generic "printableArray" function?  So you can have all future arrays use the same thing?

**Comment by tabrindle** [comment]
np - I have npm scripts for lint/format - I think Ill integrate my other project https://github.com/tabrindle/run4staged here to prevent this kind of thing :-)

File: src/envinfo.js (hunk 1 above)
Comment (tabrindle) [reply src/envinfo.js]: yeah, this way we can add iOS in the future.

File: src/helpers.js (hunk 2 above)
Comment (tabrindle) [reply src/helpers.js]: confirmed - minor change to formatters will make this work without prettifyGetAllAndroidSDKs at all. something like:

`const version = Array.isArray(v[1]) ? v[1].join(', ') : v[1];`

**Comment by GantMan** [comment]
### Updated:

* improved formatter to handle arrays
* created new section
* handles no SDKs

image

**Comment by tabrindle** [comment]
Sorry - one more thing. Do we want to delineate the difference between build-tools and platforms?

**Comment by tabrindle** [comment]
Also - we may not need it now, but in the future if we do anything else, it seems that `sdkmanager --list` has a `--verbose` option that might be easier to parse.

**Comment by GantMan** [comment]
> Do we want to delineate the difference between build-tools and platforms?

So, SDK tools are going to be important for what is needed to develop for those.

Platform is tools for that platform, like ADB etc.  

The thing that's really important most times is the SDK, which is very locked to version numbers.  If I find that I'll need Platform, then I'll def do another PR, but SDKs were the critical.

> Also - we may not need it now, but in the future if we do anything else, it seems that sdkmanager --list has a --verbose option that might be easier to parse.

`--verbose` looks like it might be a bit harder to parse, to me.  It looks like it has a lot more noise.


TL:DR; -> I think we're good as is.

**Comment by tabrindle** [comment]
I don't mean platform-tools, that is indeed different.

I mean the confusion that might stem from things like in gradle files:

android {
    compileSdkVersion 25
    buildToolsVersion "25.0.2"
}

We are returning "Android SDK versions" that are actually build-tools. 

I am referring to platforms, like in `sdkmanager --list`:

platforms;android-26

See what I mean? You can in fact have platforms installed without build-tools, and vice versa. 

Really, this is google's fault for sucking at naming and versioning, but I'd like to mitigate any confusion.

**Comment by GantMan** [comment]
I see what you're saying.   Do you want a separate function? 

Where do you want build tools printed?  New section?  You'll never use this section for any other shitty google tech, so I'm willing to add what you outline.

**Comment by GantMan** [comment]
Android madness unlocked

image

**Comment by tabrindle** [comment]
Updated so that table, markdown and json outputs all match:

### SDKs:
  #### Android:
    * Build Tools: 27.0.3
    * API Levels: 26

```
  "SDKs": {
    "Android": {
      "Build Tools": [
        "27.0.3"
      ],
      "API Levels": [
        "26"
      ]
    }
  },
```

```
SDKs:
  Android:
    Build Tools: 27.0.3
    API Levels: 26
```

Thank you so much for your help! this is awesome.

**Comment by GantMan** [comment]
Looks great @tabrindle !!!!

**Comment by tabrindle** [comment]
@GantMan So there are a few problems with the formatters, but they really aren't your fault at all. I'm working on a separate PR to fix, and as Im pretty sure you are only using the helpers in solidarity, merging as is shouldn't have any impact. 

Thanks again for all your work, and your patience!

**Comment by GantMan** [comment]
Glad to have helped!  I'll let you know how the feature fits/works on the other side as I consume the helper and use it to augment the RN plugin.
